modules/porous_flow/src/bcs/PorousFlowSinkEnthalpy.C

```diff
@@ -0,0 +1,83 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "PorousFlowSinkEnthalpy.h"
+#include "Function.h"
+#include "PorousFlowDictator.h"
+#include "SinglePhaseFluidProperties.h"
+#include "MooseVariable.h"
+
+#include "libmesh/quadrature.h"
+
+#include <iostream>
+
+registerMooseObject("PorousFlowApp", PorousFlowSinkEnthalpy);
+
+InputParameters
+PorousFlowSinkEnthalpy::validParams()
+{
+  InputParameters params = IntegratedBC::validParams();
+  params.addRequiredParam<UserObjectName>(
+      "PorousFlowDictator", "The UserObject that holds the list of PorousFlow variable names");
+  params.addRequiredParam<FunctionName>(
+      "flux_function",
+      "The mass flux.  The flux is OUT of the medium: hence positive values of "
+      "this function means this BC will act as a SINK, while negative values "
+      "indicate this flux will be a SOURCE.  The functional form is useful "
+      "for spatially or temporally varying sinks.");
+  params.addRequiredParam<UserObjectName>("fp", "The name of the user object for fluid properties");
+  params.addRequiredCoupledVar("pressure", "Pressure");
```

Comment:
This is my main concern with this design.  Everywhere else we get "pressure" (and its derivatives) from a Material.  The reason for this is that "pressure" may not be a Variable.  Eg, when you're doing 2phase systems and you're using a (P,S) version, or Chris's crazy FluidState stuff that uses different variables.   See, eg, `PorousFlowSinkPTDefiner`

Reply:
Ok. If I get pressure from the material property, how do I get the jacobian? I think you guys compute the derivatives in the material, but that means I need to "replace" your version of the material on the side with my side material, which may not be even possible. Or do other convoluted things to work around this problem. I think that's why I did it via coupled variable, because it was way easier.

Also, I don't think @rpodgorney plans to use this with other than (p, T) formulation. But, if we want to generalize it, I would need your help with  - I don't understand all of the porous flow stuff. So, if you have suggestions how I should code this up, let me know. I am happy do to the work...

Reply:
We also calculate the derivatives you need:
```
 _dpp_dvar(getMaterialProperty<std::vector<std::vector<Real>>>("dPorousFlow_porepressure_nodal_dvar")
```
then use the chain rule with your dh_dp, eg
```
jac = dh_dp * dpp_dvar[i][ph][pvar]
```
(with phi and test too), where
```
 const unsigned int pvar = _dictator.porousFlowVariableNum(jvar);
```

Reply:
and i is the node and ph is the phase (user input)

Reply:
so if `jvar` corresponds to the temperature variable, is `dpp_dvar[i][ph][pvar] == 0`? (assuming `pvar = _dictator.porousFlowVariableNum(jvar);`)

Reply:
and then, I am assuming `dpp_dvar` is going to be 1 for my case with single pressure...

Reply:
> so if `jvar` corresponds to the temperature variable, is `dpp_dvar[i][ph][pvar] == 0`? (assuming `pvar = _dictator.porousFlowVariableNum(jvar);`)

I *think* this is true for all cases currently, yes.  I have a vague feeling that we were playing with entropy being a variable, and maybe people are even using that, in which case the derivative would be nonzero.

Reply:
> and then, I am assuming `dpp_dvar` is going to be 1 for my case with single pressure...

Yes